src/libsystemd-network/sd-dhcp-lease.c

```diff
@@ -559,6 +573,123 @@ static int lease_parse_sip_server(const uint8_t *option, size_t len, struct in_a
         return lease_parse_in_addrs(option + 1, len - 1, ret, n_ret);
 }
 
+static int lease_parse_dns_name(const uint8_t *optval, size_t optlen, char **ret) {
+        _cleanup_free_ char *name = NULL;
+        int r;
+
+        assert(optval);
+        assert(ret);
+
+        r = dns_name_from_wire_format(&optval, &optlen, &name);
+        if (r < 0)
+                return r;
+        if (r == 0 || optlen != 0)
+                return -EBADMSG;
+
+        *ret = TAKE_PTR(name);
+        return r;
+}
+
+static int lease_parse_dnr(const uint8_t *option, size_t len, sd_dns_resolver **dnr, size_t *n_dnr) {
+        int r;
+        sd_dns_resolver *res_list = NULL;
+        size_t n_resolvers = 0;
+        CLEANUP_ARRAY(res_list, n_resolvers, sd_dns_resolver_array_free);
+
+        assert(option);
+        assert(dnr);
+
+        _cleanup_(sd_dns_resolver_done) sd_dns_resolver res = {};
+
+        size_t offset = 0;
+        while (offset < len) {
+                if (offset + 2 > len)
+                        return -EBADMSG;
+
+                /* Instance Data length */
+                size_t ilen = unaligned_read_be16(&option[offset]);
+                if (offset + ilen + 2 > len)
+                        return -EBADMSG;
+                offset += 2;
+                size_t iend = offset + ilen;
+
+                res.priority = unaligned_read_be16(&option[offset]);
+                offset += 2;
+
+                /* Authenticated Domain Name */
+                ilen = option[offset++];
+                if (offset + ilen > iend)
+                        return -EBADMSG;
+
+                r = lease_parse_dns_name(&option[offset], ilen, &res.auth_name);
```

Comment:
Looks like there is a memory leak somewhere:
```sh
=================================================================                                                                                             
==46366==ERROR: LeakSanitizer: detected memory leaks                                                                                                          
                                                                                                                                                              
Direct leak of 506 byte(s) in 1 object(s) allocated from:                                                                                                     
    #0 0x4f7d59 in __interceptor_realloc (/home/vagrant/systemd-asan/fuzz-build/fuzz-dhcp-client+0x4f7d59) (BuildId: ea53fd510423e379f423682a2990164aabaa63ae)
    #1 0x7ff8483b0f46 in greedy_realloc /home/vagrant/systemd-asan/fuzz-build/../src/basic/alloc-util.c:70:13                                                 
    #2 0x7ff847fef1c0 in dns_name_from_wire_format /home/vagrant/systemd-asan/fuzz-build/../src/shared/dns-domain.c:945:22
    #3 0x576b8a in lease_parse_dns_name /home/vagrant/systemd-asan/fuzz-build/../src/libsystemd-network/sd-dhcp-lease.c:583:13
    #4 0x576b8a in lease_parse_dnr /home/vagrant/systemd-asan/fuzz-build/../src/libsystemd-network/sd-dhcp-lease.c:624:21
    #5 0x56fd6a in dhcp_lease_parse_options /home/vagrant/systemd-asan/fuzz-build/../src/libsystemd-network/sd-dhcp-lease.c:1014:21
    #6 0x55c707 in parse_options /home/vagrant/systemd-asan/fuzz-build/../src/libsystemd-network/dhcp-option.c:347:33
    #7 0x55beb7 in dhcp_option_parse /home/vagrant/systemd-asan/fuzz-build/../src/libsystemd-network/dhcp-option.c:380:21
    #8 0x5560ef in client_parse_message /home/vagrant/systemd-asan/fuzz-build/../src/libsystemd-network/sd-dhcp-client.c:1477:13
    #9 0x542808 in client_handle_offer_or_rapid_ack /home/vagrant/systemd-asan/fuzz-build/../src/libsystemd-network/sd-dhcp-client.c:1562:13
    #10 0x541ed5 in LLVMFuzzerTestOneInput /home/vagrant/systemd-asan/fuzz-build/../src/libsystemd-network/fuzz-dhcp-client.c:85:13
    #11 0x442b84 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/vagrant/systemd-asan/fuzz-build/fuzz-dhcp-client+0x442b84) (BuildId: ea53fd510423e379f423682a2990164aabaa63ae)
    #12 0x441fa9 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/home/vagrant/systemd-asan/fuzz-build/fuzz-dhcp-client+0x441fa9) (BuildId: ea53fd510423e379f423682a2990164aabaa63ae)
    #13 0x4435c5 in fuzzer::Fuzzer::MutateAndTestOne() (/home/vagrant/systemd-asan/fuzz-build/fuzz-dhcp-client+0x4435c5) (BuildId: ea53fd510423e379f423682a2990164aabaa63ae)
    #14 0x443f45 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) (/home/vagrant/systemd-asan/fuzz-build/fuzz-dhcp-client+0x443f45) (BuildId: ea53fd510423e379f423682a2990164aabaa63ae)
    #15 0x430741 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd-asan/fuzz-build/fuzz-dhcp-client+0x430741) (BuildId: ea53fd510423e379f423682a2990164aabaa63ae)
    #16 0x45cf06 in main (/home/vagrant/systemd-asan/fuzz-build/fuzz-dhcp-client+0x45cf06) (BuildId: ea53fd510423e379f423682a2990164aabaa63ae)
    #17 0x7ff847249b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #18 0x7ff847249c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: 70e92bb237883be3065a6afc9f0696aef2d068bf)
    #19 0x425164 in _start (/home/vagrant/systemd-asan/fuzz-build/fuzz-dhcp-client+0x425164) (BuildId: ea53fd510423e379f423682a2990164aabaa63ae)

SUMMARY: AddressSanitizer: 506 byte(s) leaked in 1 allocation(s).
INFO: to ignore leaks on libFuzzer side use -detect_leaks=0.
```
As far as I can see the option was
```
          "dhcp.option.type": "162",
          "dhcp.option.type_tree": {
            "dhcp.option.length": "54",
            "dhcp.option.value": "00:19:00:01:09:07:fb:00:00:90:61:62:61:00:0c:00:0b:00:00:05:03:00:00:65:00:3a:00:00:00:02:64:20:38:8a:68:b0:00:f3:00:00:00:00:f3:00:00:00:76:02:64:20:38:68:b0:00"
          },
```

Reply:
Turns out that packet contained the overload DHCP option so the OPTION_V4_DNR options were stuffed into the BOOTP part. There were two DNR options there:
```python
b'\x00\x19\x00\x01\t\x07\x00\x00\x00\x03aba\x00\x0c\x00\x00\x04\x00\x00e\x00\x00\xa26\x00\x19\x00\x00\x00\x00\x00 8h\xb0\x00\x00\x003\x00\x00\x00\xa27\x00\x19\x00\x00\x00J\x00\x00z'
```
and
```python
b'\x00\x19\x00\x01\t\x07\xfb\x00\x00\x90aba\x00\x0c\x00\x0b\x00\x00\x05\x03\x00\x00e\x00:\x00\x00\x00\x02d 8\x8ah\xb0\x00\xf3\x00\x00\x00\x00\xf3\x00\x00\x00v\x02d 8h\xb0\x00'
```

Reply:
Interesting. I'm not sure how this leaks, I'll have to check.

Reply:
The memory leak can be reproduced by running something like
```sh
cat <<'EOL' | xxd -ps -r >dnr-leak
000200000e065eff2703000f0000000577480101c0d71800000101000000
00000000000000000000000000000000000e003b040000000000a2370019
0001090700000003616261000c000004000065003a000000000033000000
a23837001900000000000000000000000e003b040000000000a236001900
01090700000003616261000c0000040000650000a2360019000000000020
3868b000000033000000a23700190000004a00007a0000000000a2360019
00010907fb000090616261000c000b00000503000065003a000000026420
388a68b000f300000000f3000000760264203868b000f300000000000000
000000000000340123
EOL
```
and then passing `dnr-leak` to the fuzz target

Reply:
Ah, thanks. There was a another continue I needed to touch up following the change from linked list to array.